File: src/lex/templates/lex.js
function $prefix$Lex(text)
{
  this._text = text;
  this._index = 0;

  this.error = false;
  this.eof = this._index >= this._text.length;
}

$prefix$Lex.prototype.dfa = [
$js-state-table$];

$prefix$Lex.prototype.getNextLexeme = function()
{
  if (this._index >= this._text.length)
  {
    this.eof = true;
    return null;
  }

  var state = this.dfa[0];
  var match = null;
  var start = this._index;
  var end = this._index;

  if (state.answer != undefined)
  {
    match = {
      type: state.answer,
      end: end,
    };
  }

  while (end < this._text.length)
  {
    var letter = this._text[end++];

    if (state[letter] == undefined)
      break;

    state = this.dfa[state[letter]];

    if (state.answer != undefined)
    {
      match = {
        type: state.answer,
        end: end,
      };
    }
  }

  if (match != null)
  {
    this._index = match.end;
    this.eof = this._index >= this._text.length;

    return {
      type: match.type,
      lexeme: this._text.substr(start, match.end - start)
    };
  }
  else
  {
    this.error = true;
    return null;
  }
};

File: src/lex/templates/lex.c
#include <string.h>

typedef struct
{
  const char *text;
  unsigned long index;
  unsigned long length;

  int eof;
  int error;
} $prefix$lexer;

enum
{
$c-token-enum$};

char $prefix$lexer_table[][256] = {
$c-state-table$};
long $prefix$lexer_answer[] = { $c-answer-table$ };

void $prefix$lexer_init($prefix$lexer *lex, const char *text)
{
  lex->text = text;
  lex->index = 0;
  lex->length = strlen(text);

  lex->eof = lex->index >= lex->length;
  lex->error = 0;
}

typedef struct
{
  long        type;
  const char      *lexeme;
  unsigned long     lexeme_length;
} $prefix$lexeme;

int $prefix$lexer_next_lexeme($prefix$lexer *lex, $prefix$lexeme *out)
{
  if (lex->index >= lex->length)
  {
    lex->eof = 1;

    $prefix$lexeme res = { 0, 0, 0 };
      *out = res;

      return 0;
  }

  unsigned long i = 0;
  unsigned long start = lex->index;
  unsigned long end = lex->index;

  long      match_type = -1;
  unsigned long   match_end = -1;

  if ($prefix$lexer_answer[i] != -1)
  {
    match_type = $prefix$lexer_answer[i];
    match_end = end;
  }

  while (end < lex->length)
  {
    char letter = lex->text[end++];

    if ($prefix$lexer_table[i][letter] == -1)
      break;

    i = $prefix$lexer_table[i][letter];

    if ($prefix$lexer_answer[i] != -1)
    {
      match_type = $prefix$lexer_answer[i];
      match_end = end;
    }
  }

  if (match_type != -1)
  {
    lex->index = match_end;
    lex->eof = lex->index >= lex->length;

    $prefix$lexeme res = {
      match_type,
      lex->text + start,
      match_end - start
    };
    *out = res;

      return 1;
  }
  else
  {
    lex->error = 1;

    $prefix$lexeme res = { 0, 0, 0 };
      *out = res;

    return 0;
  }
}
